R/course.R

```diff
@@ -259,6 +264,81 @@ tidy_download <- function(url, destdir = getwd()) {
   invisible(full_path)
 }
 
+download_url <- function(url,
+                         destfile,
+                         handle = curl::new_handle(),
+                         n_tries = 3,
+                         retry_connecttimeout = 40L) {
+  handle_options <- list(noprogress = FALSE, progressfunction = progress_fun)
+  curl::handle_setopt(handle, .list = handle_options)
+
+  we_should_retry <- function(i, n_tries, status) {
+    if (i >= n_tries) {
+      FALSE
+    } else if (inherits(status, "error")) {
+      # TODO: find a way to detect a (connect) timeout more specifically?
+      # https://github.com/jeroen/curl/issues/154
+      # https://ec.haxx.se/usingcurl/usingcurl-timeouts
+      # "Failing to connect within the given time will cause curl to exit with a
+      # timeout exit code (28)."
+      # (however, note that all timeouts lead to this same exit code)
+      # https://ec.haxx.se/usingcurl/usingcurl-returns
+      # "28. Operation timeout. The specified time-out period was reached
+      # according to the conditions. curl offers several timeouts, and this exit
+      # code tells one of those timeout limits were reached."
+      # https://github.com/curl/curl/blob/272282a05416e42d2cc4a847a31fd457bc6cc827/lib/strerror.c#L143-L144
+      # "Timeout was reached" <-- actual message we could potentially match
```

Comment:
Ideally we would act, knowing what sort of error we got. But I think that's not possible today.

Silver living: I'm sure there are intermittent failures for reasons other than connect timeout, so this does give us a general way to retry a couple of times, for whatever reason.